cpp/src/arrow/util/align_util.cc

```diff
@@ -0,0 +1,215 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/util/align_util.h"
+
+#include "arrow/array.h"
+#include "arrow/chunked_array.h"
+#include "arrow/record_batch.h"
+#include "arrow/table.h"
+
+namespace arrow {
+
+namespace util {
+
+bool CheckAlignment(const Buffer& buffer, const int64_t& alignment) {
+  return buffer.address() % alignment == 0;
+}
+
+bool CheckAlignment(const ArrayData& array, const int64_t& alignment) {
+  for (const auto& buffer : array.buffers) {
+    if (buffer) {
+      if (!CheckAlignment(*buffer, alignment)) return false;
+    }
+  }
+  return true;
+}
+
+bool CheckAlignment(const Array& array, const int64_t& alignment) {
+  bool align_dictionary = true, align_children = true;
+
+  if (array.type()->id() == Type::DICTIONARY) {
+    align_dictionary = CheckAlignment(*array.data()->dictionary, alignment);
+  }
+
+  for (const auto& child : array.data()->child_data) {
+    if (child) {
+      if (!CheckAlignment(*child, alignment)) align_children = false;
+    }
+  }
+  return CheckAlignment(*array.data(), alignment) && align_dictionary && align_children;
+}
+
+bool CheckAlignment(const ChunkedArray& array, const int64_t& alignment,
+                    std::vector<bool>& needs_alignment, const int& offset) {
+  needs_alignment.resize(needs_alignment.size() + array.num_chunks(), false);
+  for (auto i = 0; i < array.num_chunks(); ++i) {
+    if (array.chunk(i) && !CheckAlignment(*array.chunk(i), alignment))
+      needs_alignment[i + offset] = true;
+  }
+  return std::find(needs_alignment.begin() + offset,
```

Comment:
Instead of doing `std::find` at the end (which will loop through `needs_alignment`) could you do something like...

```
bool at_least_one_needs_alignment = false;
for (...) {
  if (...) {
    needs_alignment[i + offset] = true;
    at_least_one_needs_alignment = true;
  }
}
return at_least_one_needs_alignment;
```